lib/api/bucketPut.js

```diff
@@ -39,36 +56,79 @@ export default function bucketPut(authInfo, request, locationConstraint, log,
+                        awsService: 's3',
+                        locationConstraint,
+                        requesterInfo: authInfo,
+                        signatureVersion: authParams.params.data.authType,
+                        authType: authParams.params.data.signatureVersion,
+                        signatureAge: authParams.params.data.signatureAge,
+                    },
+                };
+                return vault.checkPolicies(requestContextParams,
+                    authInfo.getArn(), log, (err, authorizationResults) => {
+                        if (err) {
+                            return next(err);
+                        }
+                        if (authorizationResults[0].isAllowed !== true) {
+                            log.trace('authorization check failed for user',
+                                { locationConstraint });
+                            return next(errors.AccessDenied);
+                        }
+                        return next(null, locationConstraint);
+                    });
             }
-            pushMetric('createBucket', log, {
-                authInfo,
-                bucket: bucketName,
-            });
-            return callback(null, corsHeaders);
-        });
+            return next(null, locationConstraint);
+        },
+        // AWS JS SDK sends a request with locationConstraint us-east-1 if no
+        // locationConstraint provided.
+        (locationConstraint, next) => {
+            const { locationConstraints, restEndpoints } = config;
+            const { parsedHost } = request;
+            if (locationConstraint && locationConstraints &&
+                !Object.keys(locationConstraints)
+                    .includes(locationConstraint)) {
+                log.trace('locationConstraint is invalid',
+                    { locationConstraint });
+                return next(errors.InvalidLocationConstraint);
+            }
+            if (!locationConstraint && restEndpoints && parsedHost &&
+                restEndpoints[parsedHost]) {
+                return next(null, restEndpoints[parsedHost]);
+            }
+            return next(null, locationConstraint);
+        },
+        (locationConstraint, next) => createBucket(authInfo, bucketName,
+            request.headers, locationConstraint, config.usEastBehavior, log,
+            (err, previousBucket) => {
+                // if bucket already existed, gather any relevant cors headers
+                const corsHeaders = collectCorsHeaders(
+                    request.headers.origin, request.method, previousBucket);
+                if (err) {
+                    return next(err, corsHeaders);
+                }
+                pushMetric('createBucket', log, {
+                    authInfo,
+                    bucket: bucketName,
+                });
+                return next(null, corsHeaders);
+            }),
+    ], (err, corsHeaders) => callback(err, corsHeaders));
```

Comment:
`(err, corsHeaders) => callback(err, corsHeaders)` need not be expanded, sending callback as param is sufficient.